File: web-api/providers/gcp.mjs
import crypto from "node:crypto";
import { decryptSecret } from "../core/crypto.mjs";
import { getAccountSecretRecord } from "../repositories/accounts.mjs";

export function meta(row) {
  let value = {};
  try { value = JSON.parse(row.credential_meta || "{}"); } catch { /* legacy account */ }
  const projectId = String(value.project_id || "").trim();
  if (!projectId) throw new Error("GCP 账号缺少 Project ID");
  return { projectId };
}

function base64Url(value) { return Buffer.from(value).toString("base64url"); }

export async function token(accountId) {
  const row = getAccountSecretRecord(accountId);
  if (!row) throw new Error("云账号不存在");
  if (!row.enabled) throw new Error("云账号已停用");
  if (row.cloud_type !== "gcp") throw new Error("当前账号不是 Google Cloud 账号");
  const { projectId } = meta(row);
  const now = Math.floor(Date.now() / 1000);
  const header = base64Url(JSON.stringify({ alg: "RS256", typ: "JWT" }));
  const claim = base64Url(JSON.stringify({ iss: row.access_key_id, scope: "https://www.googleapis.com/auth/cloud-platform", aud: "https://oauth2.googleapis.com/token", iat: now, exp: now + 3600 }));
  let signature;
  try { signature = crypto.sign("RSA-SHA256", Buffer.from(`${header}.${claim}`), decryptSecret(row.secret_ciphertext).replace(/\\n/g, "\n")).toString("base64url"); }
  catch { throw new Error("GCP 服务账号私钥无效，需填写未加密的 PEM 私钥"); }
  const body = new URLSearchParams({ grant_type: "urn:ietf:params:oauth:grant-type:jwt-bearer", assertion: `${header}.${claim}.${signature}` });
  const response = await fetch("https://oauth2.googleapis.com/token", { method: "POST", headers: { "Content-Type": "application/x-www-form-urlencoded" }, body });
  const data = await response.json().catch(() => ({}));
  if (!response.ok || !data.access_token) throw new Error(data.error_description || data.error || `GCP OAuth ${response.status}`);
  return { token: data.access_token, projectId, accountId: row.access_key_id };
}

export async function get(context, url) {
  const response = await fetch(url, { headers: { Authorization: `Bearer ${context.token}` } });
  const data = await response.json().catch(() => ({}));
  if (!response.ok) throw new Error(data.error?.message || data.message || `GCP ${response.status}`);
  return data;
}

export async function pages(context, url, key) {
  const items = [];
  let next = url;
  for (let index = 0; next && index < 100; index += 1) {
    const data = await get(context, next);
    items.push(...(Array.isArray(data[key]) ? data[key] : []));
    next = data.nextPageToken ? `${url}${url.includes("?") ? "&" : "?"}pageToken=${encodeURIComponent(data.nextPageToken)}` : "";
  }
  return items;
}

function instance(item, region) { const network = item.networkInterfaces?.[0] || {}; return { ...item, InstanceId: String(item.id || item.name), InstanceName: item.name, InstanceStatus: item.status, Status: item.status, PublicIpAddress: network.accessConfigs?.[0]?.natIP || "", PrivateIpAddress: network.networkIP || "", InstanceType: String(item.machineType || "").split("/").pop() || "", VpcId: String(network.network || "").split("/").pop() || "", _region_id: region }; }
function rds(item) { const settings = item.settings || {}; return { ...item, DBInstanceId: item.name, DBInstanceDescription: item.name, DBInstanceStatus: item.state, DBInstanceClass: settings.tier || "", DBInstanceStorage: Number(settings.dataDiskSizeGb || 0), ConnectionString: item.ipAddresses?.find((ip) => ip.type === "PRIMARY")?.ipAddress || "", Port: "3306", Engine: item.databaseVersion || "", EngineVersion: item.databaseVersion || "", CreateTime: item.createTime || "", _region_id: item.region || "" }; }
function redis(item) { const size = Number(item.memorySizeGb || 0); return { ...item, InstanceId: item.name, InstanceName: String(item.name || "").split("/").pop(), InstanceStatus: item.state, InstanceType: "Redis", InstanceClass: item.tier || "", Capacity: size * 1024, ConnectionDomain: item.host || "", Port: item.port || "", EngineVersion: item.redisVersion || "", NetworkType: item.authorizedNetwork || "", _region_id: item.locationId || "" }; }
function bucket(item) { return { ...item, Name: item.name, BucketName: item.name, Location: item.location || "", CreationDate: item.timeCreated || "", StorageClass: item.storageClass || "STANDARD", Acl: item.iamConfiguration?.uniformBucketLevelAccess?.enabled ? "private" : "unknown", ExtranetEndpoint: "-", IntranetEndpoint: "-", _region_id: item.location || "" }; }
function zone(item) { return { ...item, DomainName: String(item.dnsName || "").replace(/\.$/, ""), DomainStatus: "ACTIVE", ZoneId: item.id || item.name, RecordCount: Number(item?.cloudLoggingConfig ? 0 : 0), RegistrationDate: item.creationTime || "", _region_id: "global", _gcp_dns: true, ...item }; }

export async function resources(accountId, type) {
  const context = await token(accountId);
  const project = encodeURIComponent(context.projectId);
  try {
    if (type === "ecs") {
      const data = await get(context, `https://compute.googleapis.com/compute/v1/projects/${project}/aggregated/instances`);
      const items = Object.entries(data.items || {}).flatMap(([scope, value]) => (value?.instances || []).map((item) => instance(item, scope.split("/").pop() || "")));
      return { resource_type: type, items, errors: [], fetched_at: Date.now() };
    }
    const definitions = { rds: [`https://sqladmin.googleapis.com/sql/v1beta4/projects/${project}/instances`, "items", rds], redis: [`https://redis.googleapis.com/v1/projects/${project}/locations/-/instances`, "instances", redis], oss: [`https://storage.googleapis.com/storage/v1/b?project=${project}`, "items", bucket], domain: [`https://dns.googleapis.com/dns/v1/projects/${project}/managedZones`, "managedZones", zone] };
    const definition = definitions[type];
    if (!definition) return { resource_type: type, items: [], errors: [`GCP 暂未接入 ${type} 资源`], fetched_at: Date.now() };
    const [url, key, normalize] = definition;
    return { resource_type: type, items: (await pages(context, url, key)).map(normalize), errors: [], fetched_at: Date.now() };
  } catch (error) { return { resource_type: type, items: [], errors: [error.message], fetched_at: Date.now() }; }
}

export async function verify(accountId, regions = ["asia-east1"]) {
  const context = await token(accountId);
  await get(context, `https://cloudresourcemanager.googleapis.com/v1/projects/${encodeURIComponent(context.projectId)}`);
  return { provider: "gcp", verified: true, region_count: regions.length, regions, default_region: regions[0] };
}

File: src-tauri/src/cloud/gcp.rs
use chrono::Utc;
use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine};
use rsa::{pkcs1::DecodeRsaPrivateKey, pkcs8::DecodePrivateKey, RsaPrivateKey};
use rsa::pkcs1v15::SigningKey;
use rsa::signature::{SignatureEncoding, Signer};
use serde_json::{json, Value};
use sha2::Sha256;
use crate::ResourceResponse;

struct GcpCredentials { email: String, private_key: String, project_id: String }

fn credentials(id: i64) -> Result<GcpCredentials, String> {
    let conn = crate::open_db()?;
    let (email, secret_ciphertext, enabled) = crate::account_repository::credential_record(&conn, id)?;
    if enabled != 1 { return Err("云账号已停用".into()); }
    if crate::account_repository::cloud_type(&conn, id)? != "gcp" { return Err("当前账号不是 Google Cloud 账号".into()); }
    let project_id = crate::account_repository::credential_meta(&conn, id)?.get("project_id").and_then(Value::as_str).unwrap_or("").trim().to_string();
    if project_id.is_empty() { return Err("GCP 账号缺少 Project ID".into()); }
    Ok(GcpCredentials { email, private_key: crate::decrypt_secret(&secret_ciphertext)?.replace("\\n", "\n"), project_id })
}

async fn token(credentials: &GcpCredentials) -> Result<String, String> {
    let header = URL_SAFE_NO_PAD.encode(r#"{"alg":"RS256","typ":"JWT"}"#); let now = Utc::now().timestamp(); let claim = URL_SAFE_NO_PAD.encode(serde_json::to_string(&json!({"iss": credentials.email, "scope": "https://www.googleapis.com/auth/cloud-platform", "aud": "https://oauth2.googleapis.com/token", "iat": now, "exp": now + 3600})).map_err(|error| error.to_string())?); let signing_input = format!("{header}.{claim}"); let private_key = RsaPrivateKey::from_pkcs8_pem(&credentials.private_key).or_else(|_| RsaPrivateKey::from_pkcs1_pem(&credentials.private_key)).map_err(|_| "GCP 服务账号私钥无效，需使用未加密的 PEM 私钥".to_string())?; let signature = URL_SAFE_NO_PAD.encode(SigningKey::<Sha256>::new(private_key).sign(signing_input.as_bytes()).to_vec()); let assertion = format!("{signing_input}.{signature}");
    let response = reqwest::Client::new().post("https://oauth2.googleapis.com/token").form(&[("grant_type", "urn:ietf:params:oauth:grant-type:jwt-bearer"), ("assertion", assertion.as_str())]).timeout(std::time::Duration::from_secs(30)).send().await.map_err(|error| format!("GCP OAuth 请求失败: {error}"))?; let status = response.status(); let data: Value = response.json().await.map_err(|error| format!("GCP OAuth 返回解析失败: {error}"))?; if !status.is_success() { return Err(data.get("error_description").or_else(|| data.get("error")).and_then(Value::as_str).unwrap_or("GCP OAuth 失败").to_string()); } data.get("access_token").and_then(Value::as_str).map(String::from).ok_or_else(|| "GCP OAuth 未返回 access token".into())
}

async fn get(token: &str, url: &str) -> Result<Value, String> { let response = reqwest::Client::new().get(url).bearer_auth(token).timeout(std::time::Duration::from_secs(30)).send().await.map_err(|error| format!("GCP 请求失败: {error}"))?; let status = response.status(); let data: Value = response.json().await.map_err(|error| format!("GCP 返回解析失败: {error}"))?; if !status.is_success() { return Err(data.pointer("/error/message").or_else(|| data.get("message")).and_then(Value::as_str).unwrap_or("GCP API 返回错误").to_string()); } Ok(data) }

async fn pages(token: &str, url: &str, key: &str) -> Result<Vec<Value>, String> { let mut items = Vec::new(); let mut next = url.to_string(); for _ in 0..100 { let data = get(token, &next).await?; items.extend(crate::array_at(&data, &[key]).into_iter().cloned()); let Some(page_token) = data.get("nextPageToken").and_then(Value::as_str) else { break; }; next = format!("{}{}pageToken={}", url, if url.contains('?') { "&" } else { "?" }, crate::rpc_encode(page_token)); } Ok(items) }

fn instance(item: &Value, region: &str) -> Value { let network = item.pointer("/networkInterfaces/0"); json!({"InstanceId": item.get("id").or_else(|| item.get("name")), "InstanceName": item.get("name"), "InstanceStatus": item.get("status"), "Status": item.get("status"), "PublicIpAddress": network.and_then(|v| v.pointer("/accessConfigs/0/natIP")), "PrivateIpAddress": network.and_then(|v| v.get("networkIP")), "InstanceType": item.get("machineType").and_then(Value::as_str).unwrap_or("").rsplit('/').next().unwrap_or(""), "VpcId": network.and_then(|v| v.get("network")).and_then(Value::as_str).unwrap_or("").rsplit('/').next().unwrap_or(""), "_region_id": region, "_raw": item}) }
fn rds(item: &Value) -> Value { json!({"DBInstanceId": item.get("name"), "DBInstanceDescription": item.get("name"), "DBInstanceStatus": item.get("state"), "DBInstanceClass": item.pointer("/settings/tier"), "DBInstanceStorage": item.pointer("/settings/dataDiskSizeGb").unwrap_or(&json!(0)), "ConnectionString": item.get("ipAddresses").and_then(Value::as_array).and_then(|v| v.iter().find(|x| x.get("type").and_then(Value::as_str) == Some("PRIMARY"))).and_then(|v| v.get("ipAddress")), "Port": "3306", "Engine": item.get("databaseVersion"), "EngineVersion": item.get("databaseVersion"), "CreateTime": item.get("createTime"), "_region_id": item.get("region"), "_raw": item}) }
fn redis(item: &Value) -> Value { json!({"InstanceId": item.get("name"), "InstanceName": item.get("name").and_then(Value::as_str).unwrap_or("").rsplit('/').next().unwrap_or(""), "InstanceStatus": item.get("state"), "InstanceType": "Redis", "InstanceClass": item.get("tier"), "Capacity": item.get("memorySizeGb").and_then(Value::as_i64).unwrap_or(0) * 1024, "ConnectionDomain": item.get("host"), "Port": item.get("port"), "EngineVersion": item.get("redisVersion"), "NetworkType": item.get("authorizedNetwork"), "_region_id": item.get("locationId"), "_raw": item}) }
fn bucket(item: &Value) -> Value { json!({"Name": item.get("name"), "BucketName": item.get("name"), "Location": item.get("location"), "CreationDate": item.get("timeCreated"), "StorageClass": item.get("storageClass").unwrap_or(&json!("STANDARD")), "Acl": "private", "ExtranetEndpoint": "-", "IntranetEndpoint": "-", "_region_id": item.get("location"), "_raw": item}) }
fn zone(item: &Value) -> Value { json!({"DomainName": item.get("dnsName").and_then(Value::as_str).unwrap_or("").trim_end_matches('.'), "DomainStatus": "ACTIVE", "ZoneId": item.get("id").or_else(|| item.get("name")), "RecordCount": 0, "RegistrationDate": item.get("creationTime"), "_region_id": "global", "_gcp_dns": true, "_raw": item}) }

pub(crate) async fn resource_items(id: i64, resource_type: &str) -> ResourceResponse {
    let now = Utc::now().timestamp_millis(); let credentials = match credentials(id) { Ok(v) => v, Err(e) => return ResourceResponse { resource_type: resource_type.into(), items: vec![], errors: vec![e], fetched_at: now } }; let token = match token(&credentials).await { Ok(v) => v, Err(e) => return ResourceResponse { resource_type: resource_type.into(), items: vec![], errors: vec![e], fetched_at: now } }; let project = crate::rpc_encode(&credentials.project_id);
    let result = match resource_type {
        "ecs" => get(&token, &format!("https://compute.googleapis.com/compute/v1/projects/{project}/aggregated/instances")).await.map(|data| {
            data.get("items").and_then(Value::as_object).into_iter().flat_map(|values| values.iter()).flat_map(|(scope, value)| {
                crate::array_at(value, &["instances"]).into_iter().map(|item| instance(item, scope.rsplit('/').next().unwrap_or(""))).collect::<Vec<_>>()
            }).collect::<Vec<_>>()
        }),
        "rds" => pages(&token, &format!("https://sqladmin.googleapis.com/sql/v1beta4/projects/{project}/instances"), "items").await.map(|v| v.into_iter().map(|item| rds(&item)).collect()),
        "redis" => pages(&token, &format!("https://redis.googleapis.com/v1/projects/{project}/locations/-/instances"), "instances").await.map(|v| v.into_iter().map(|item| redis(&item)).collect()),
        "oss" => pages(&token, &format!("https://storage.googleapis.com/storage/v1/b?project={project}"), "items").await.map(|v| v.into_iter().map(|item| bucket(&item)).collect()),
        "domain" => pages(&token, &format!("https://dns.googleapis.com/dns/v1/projects/{project}/managedZones"), "managedZones").await.map(|v| v.into_iter().map(|item| zone(&item)).collect()),
        _ => return ResourceResponse { resource_type: resource_type.into(), items: vec![], errors: vec![format!("GCP 暂未接入 {resource_type} 资源")], fetched_at: now }
    };
    match result { Ok(items) => ResourceResponse { resource_type: resource_type.into(), items, errors: vec![], fetched_at: now }, Err(error) => ResourceResponse { resource_type: resource_type.into(), items: vec![], errors: vec![error], fetched_at: now } }
}

pub(crate) async fn verify_account(id: i64) -> Result<Value, String> {
    let credentials = credentials(id)?;
    let token = token(&credentials).await?;
    get(&token, &format!("https://cloudresourcemanager.googleapis.com/v1/projects/{}", crate::rpc_encode(&credentials.project_id))).await?;
    let regions = crate::configured_regions(id, "asia-east1")?;
    let default_region = regions.first().cloned().ok_or_else(|| "GCP 未配置可用区域".to_string())?;
    Ok(json!({"provider":"gcp","verified":true,"region_count":regions.len(),"regions":regions,"default_region":default_region}))
}
